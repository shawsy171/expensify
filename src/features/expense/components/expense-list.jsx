// @ts-check
import React from 'react';

// store
import { connect } from 'react-redux';

const ExpenseList = (props) => {
  const mapList = () => (
    props.expenses.map((expense, index) => <li key={index}>{expense.description}</li>)
  )
  
  return (
    <div>
      <h1>Expense List</h1>
      { props.filters.text }
      { mapList() }
    </div>
  )
}

const mapStateToProps = (state) => (
  {
    expenses: state.expenses,
    filters: state.filters,
  }
)

export default connect(mapStateToProps)(ExpenseList);


