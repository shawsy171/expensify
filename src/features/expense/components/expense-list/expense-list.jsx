// @ts-check
import React from 'react';

// store
import { connect } from 'react-redux';

// components
import ExpenseListItem from './components/expense-list-item/expense-list-item.jsx';

// selectors
import selectorExpenses from './../../store/expense/selectors';

const ExpenseList = (props) => {
  const mapList = () => (
    props.expenses.map((expense, index) => 
      <ExpenseListItem key={index} expense={expense} />
    )
  )
  
  return (
    <div>
      <h1>Expense List</h1>
      { mapList() }
    </div>
  )
}

const mapStateToProps = (state) => (
  {
    expenses: selectorExpenses(state.expenses, state.filters),
  }
)

export default connect(mapStateToProps)(ExpenseList);


