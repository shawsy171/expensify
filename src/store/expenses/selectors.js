/** filters and sorts an array of expenses 
 * @param { Array<Object> } expenses 
 * @interface Array<{
 *  text: string, 
 *  sortBy: string,
 *  startDate: number,
 *  endDate: number,
 * }>
 * @example [{
 *  description = 'Rent', 
 *  note = 'to be paid',
 *  amount = 5000,
 *  createdAt = 19299303
 * }, {
 *  description = 'Coffee', 
 *  note = 'I love coffee in the morning',
 *  amount = 300,
 *  createdAt = 85849307
 * }]
 * @param { Object } filterObject
 * @interface {
 *  text?: string, 
 *  sortBy: string,
 *  startDate?: number,
 *  endDate?: number,
 * }
 * @example {
 *  text: 'rent', 
 *  sortBy: 'amount',
 *  startDate: 100,
 *  endDate: 1250,
 * }
 * @returns { Array<Object> } expenses
 */
export const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
}