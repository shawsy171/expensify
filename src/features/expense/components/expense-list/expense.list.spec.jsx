import React from 'react';
import { shallow } from 'enzyme';
import expenses from './../../../../tests/fixtures';

import { ExpenseList } from './expense-list';

describe('ExpenseList Component', () => {
  test('should render expneses correctly', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render <p>No expenses</p> when expense list is zero', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
