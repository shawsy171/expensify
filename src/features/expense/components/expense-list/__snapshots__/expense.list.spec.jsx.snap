// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList Component should render <p>No expenses</p> when expense list is zero 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No expenses
  </p>
</div>
`;

exports[`ExpenseList Component should render expneses correctly 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "gym",
        "id": "1",
        "note": "",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "rent",
        "id": "2",
        "note": "",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    key="3"
  />
</div>
`;
